File: crates/defi-protocols/src/liquid_staking/generic_lst.rs
use alloy::primitives::{Address, U256};
use alloy::sol;
use alloy::sol_types::SolCall;
use async_trait::async_trait;

use defi_core::error::{DefiError, Result};
use defi_core::traits::LiquidStaking;
use defi_core::types::*;

sol! {
    #[sol(rpc)]
    interface IGenericLST {
        function stake() external payable returns (uint256);
        function unstake(uint256 amount) external returns (uint256);
    }
}

pub struct GenericLst {
    name: String,
    staking: Address,
}

impl GenericLst {
    pub fn from_contracts(
        name: String,
        contracts: &std::collections::HashMap<String, Address>,
    ) -> Result<Self> {
        let staking = contracts
            .get("staking")
            .copied()
            .ok_or_else(|| DefiError::ContractError("Missing 'staking' contract".to_string()))?;
        Ok(Self { name, staking })
    }
}

#[async_trait]
impl LiquidStaking for GenericLst {
    fn name(&self) -> &str {
        &self.name
    }

    async fn build_stake(&self, params: StakeParams) -> Result<DeFiTx> {
        let call = IGenericLST::stakeCall {};
        Ok(DeFiTx {
            description: format!("[{}] Stake {} HYPE", self.name, params.amount),
            to: self.staking,
            data: call.abi_encode().into(),
            value: params.amount,
            gas_estimate: Some(200_000),
        })
    }

    async fn build_unstake(&self, params: UnstakeParams) -> Result<DeFiTx> {
        let call = IGenericLST::unstakeCall {
            amount: params.amount,
        };
        Ok(DeFiTx {
            description: format!("[{}] Unstake {}", self.name, params.amount),
            to: self.staking,
            data: call.abi_encode().into(),
            value: U256::ZERO,
            gas_estimate: Some(200_000),
        })
    }

    async fn get_info(&self) -> Result<StakingInfo> {
        Err(DefiError::Unsupported(format!(
            "[{}] get_info requires RPC",
            self.name
        )))
    }
}

File: ts/packages/defi-protocols/src/liquid_staking/generic_lst.ts
import { parseAbi, encodeFunctionData } from "viem";
import type { Address } from "viem";
import type { ILiquidStaking } from "@hypurrquant/defi-core";
import {
  DefiError,
  type ProtocolEntry,
  type StakeParams,
  type UnstakeParams,
  type StakingInfo,
  type DeFiTx,
} from "@hypurrquant/defi-core";

const GENERIC_LST_ABI = parseAbi([
  "function stake() external payable returns (uint256)",
  "function unstake(uint256 amount) external returns (uint256)",
]);

export class GenericLstAdapter implements ILiquidStaking {
  private readonly protocolName: string;
  private readonly staking: Address;

  constructor(entry: ProtocolEntry, _rpcUrl?: string) {
    this.protocolName = entry.name;
    const staking = entry.contracts?.["staking"];
    if (!staking) throw DefiError.contractError("Missing 'staking' contract");
    this.staking = staking;
  }

  name(): string {
    return this.protocolName;
  }

  async buildStake(params: StakeParams): Promise<DeFiTx> {
    const data = encodeFunctionData({ abi: GENERIC_LST_ABI, functionName: "stake" });
    return {
      description: `[${this.protocolName}] Stake ${params.amount} HYPE`,
      to: this.staking,
      data,
      value: params.amount,
      gas_estimate: 200_000,
    };
  }

  async buildUnstake(params: UnstakeParams): Promise<DeFiTx> {
    const data = encodeFunctionData({
      abi: GENERIC_LST_ABI,
      functionName: "unstake",
      args: [params.amount],
    });
    return {
      description: `[${this.protocolName}] Unstake ${params.amount}`,
      to: this.staking,
      data,
      value: 0n,
      gas_estimate: 200_000,
    };
  }

  async getInfo(): Promise<StakingInfo> {
    throw DefiError.unsupported(`[${this.protocolName}] getInfo requires RPC`);
  }
}
